### PR TITLE
bibtex: remove escaping from math expressions

### PR DESCRIPTION
**Pull request by MJedr**
* INSPIR-3243

### CODE REVIEW

File: backend/inspirehep/records/serializers/bibtex.py [hunk 1]
```diff
@@ -16,6 +19,24 @@
...
+        import latexcodec  # NOQA
+
+        if "$" not in text:
+            return codecs.encode(text, "ulatex+{}".format(self.encoding))
+        else:
+            found_math_expressions = [
+                f"{expression}" for expression in re.findall(r"\$.*\$", text)
+            ]
+            encoded_text = codecs.encode(text, "ulatex+{}".format(self.encoding))
+            for list_nb, _math_expression in enumerate(
```
Comment (michamos) [inline backend/inspirehep/records/serializers/bibtex.py]: * Better to use [`zip`](https://docs.python.org/3.6/library/functions.html#zip) than fiddling with list indices for this kind of things. 
* Why the leading `_`?
* Nice trick, I had something much more complex in mind (namely splitting the string on math boundaries while keeping track of which mode every part is, and only encoding those that need encoding). Although it shouldn't happen, there's the risk that you get a different number of math segments in the non-encoded and encoded strings.
I would add an assertion like
```python
assert len(found_math_expressions) == len(encoded_math_expressions), f"number of math parts in LaTeX-encoded string is different for '{text}'."
```
before to make sure this doesn't happen.

File: backend/tests/integration/records/serializers/test_bibtex.py [hunk 2]
```diff
@@ -139,3 +141,29 @@ def test_bibtex_search(api_client, db, es, create_record):
     assert expected_status_code == response_status_code
     assert expected_result_1 in response_data
     assert expected_result_2 in response_data
+
+
+def test_bibtex_with_math_environment_in_title(api_client, db, es, create_record):
```
Comment (michamos) [inline backend/tests/integration/records/serializers/test_bibtex.py]: can you add tests for the corner case I mentioned, and a test checking that text is correctly encoded? These are strings you can use:
* `"Low-energy theorem for $\\gamma\\to 3\\pi$: Σ surface terms against $\\pi a_1$-mixing"` (this one is a bit fake, I added the `Σ`)
* `"Core polarization effects up to 12ℏω in 7Li and 10B nuclei"` (this one is real but latexcodec doesn't know about these characters, so it's not encoded properly, you should give the correct expected output and xfail the test).

File: backend/tests/integration/records/serializers/test_bibtex.py [hunk 3]
```diff
@@ -139,3 +142,55 @@ def test_bibtex_search(api_client, db, es, create_record):
...
+                "title": "Low-energy theorem for $\\gamma\\to 3\\pi$: Σ surface terms against $\\pi a_1$-mixing"
+            }
+        ],
+    }
+    record = create_record("lit", data=data)
+    record_control_number = record["control_number"]
+
+    expected_status_code = 200
+    expected_etag = '"application/x-bibtex@v1"'
+    expected_result = '@article{637275237,\n    title = "{Low-energy theorem for $\\gamma\\to 3\\pi$: Σ surface terms against $\\pi a_1$-mixing}"\n}\n'
```
Comment (michamos) [inline backend/tests/integration/records/serializers/test_bibtex.py]: what I expect is
```suggestion
    expected_result = '@article{637275237,\n    title = "{Low-energy theorem for $\\gamma\\to 3\\pi$: $\Sigma$ surface terms against $\\pi a_1$-mixing}"\n}\n'
```

File: backend/inspirehep/records/serializers/bibtex.py [hunk 4]
```diff
@@ -16,6 +19,30 @@
...
+            encoded_math_expressions = re.findall(math_expression_regex, encoded_text)
+            for encoded_math_expression, found_math_expression in zip(
+                encoded_math_expressions, found_math_expressions
+            ):
+                encoded_text = encoded_text.replace(
+                    encoded_math_expression, found_math_expression
+                )
+            assert len(found_math_expressions) == len(
+                encoded_math_expressions
+            ), f"number of math parts in LaTeX-encoded string is different for '{text}'."
```
Comment (michamos) [inline backend/inspirehep/records/serializers/bibtex.py]: in general it's better to put assertions doing sanity checks before the code that uses the assumption, so
1. you don't do useless work in case it fails
2. you can rely on the assumption being true when you need it

File: backend/inspirehep/records/serializers/bibtex.py [hunk 5]
```diff
@@ -16,6 +19,30 @@
 
 
 class BibtexWriter(Writer):
+    def _encode(self, text):
+        import latexcodec  # NOQA
+
+        if "$" not in text:
+            return codecs.encode(text, "ulatex+{}".format(self.encoding))
+        else:
+            math_expression_regex = re.compile(r"(?<!\\)\$.*?\$")
```
Comment (michamos) [inline backend/inspirehep/records/serializers/bibtex.py]: it's better to put this kind of constants out of function bodies, so you don't need to evaluate it on every function call and it's more visible. Here you can put it on the class.

File: backend/inspirehep/records/serializers/bibtex.py [hunk 6]
```diff
@@ -16,6 +19,27 @@
...
+            found_math_expressions = re.findall(math_expression_regex, text)
```
Comment (michamos) [inline backend/inspirehep/records/serializers/bibtex.py]: you can also do
```suggestion
            found_math_expressions = math_expression_regex.findall(text)
```
and maybe `found_math_expressions` is a bit long, what about `math_expressions`?

File: backend/inspirehep/records/serializers/bibtex.py [hunk 7]
```diff
@@ -16,6 +19,27 @@
...
+            encoded_text = codecs.encode(text, "ulatex+{}".format(self.encoding))
+            encoded_math_expressions = re.findall(math_expression_regex, encoded_text)
```
Comment (michamos) [inline backend/inspirehep/records/serializers/bibtex.py]: same
```suggestion
            encoded_math_expressions = math_expression_regex.findall(encoded_text)
```

File: backend/inspirehep/records/serializers/bibtex.py [hunk 8]
```diff
@@ -16,6 +19,27 @@
...
+            assert len(found_math_expressions) == len(
+                encoded_math_expressions
+            ), f"number of math parts in LaTeX-encoded string is different for '{text}'."
+            for encoded_math_expression, found_math_expression in zip(
```
Comment (michamos) [inline backend/inspirehep/records/serializers/bibtex.py]: again, variable names are a bit long here I think. What about
```suggestion
            for encoded_math, original_math in zip(
```

File: backend/tests/integration/records/serializers/test_bibtex.py (hunk 3 above)
Comment (michamos) [reply backend/tests/integration/records/serializers/test_bibtex.py]: this is important, otherwise it seems you're not encoding anything.

File: backend/tests/integration/records/serializers/test_bibtex.py [hunk 9]
```diff
@@ -139,3 +142,55 @@ def test_bibtex_search(api_client, db, es, create_record):
     assert expected_status_code == response_status_code
     assert expected_result_1 in response_data
     assert expected_result_2 in response_data
+
+
+def test_bibtex_with_math_environment_in_title(api_client, db, es, create_record):
```
Comment (michamos) [inline backend/tests/integration/records/serializers/test_bibtex.py]:
```suggestion
def test_bibtex_doesnt_encode_math_environments(api_client, db, es, create_record):
```

File: backend/tests/integration/records/serializers/test_bibtex.py [hunk 10]
```diff
@@ -139,3 +142,55 @@ def test_bibtex_search(api_client, db, es, create_record):
...
+    )
+
+    response_status_code = response.status_code
+    response_data = response.get_data(as_text=True)
+    assert expected_status_code == response_status_code
+    assert expected_result == response_data
+
+
+@pytest.mark.xfail(reason="latexcodec doesn't know about the special characters")
+def test_bibtex_with_not_encoded_math_environment_in_title_fails(
```
Comment (michamos) [inline backend/tests/integration/records/serializers/test_bibtex.py]:
```suggestion
def test_bibtex_encodes_unicode_outside_of_math_environments(
```

File: backend/tests/integration/records/serializers/test_bibtex.py [hunk 11]
```diff
@@ -139,3 +142,55 @@ def test_bibtex_search(api_client, db, es, create_record):
...
+        "control_number": 637275237,
+        "titles": [
+            {"title": "Core polarization effects up to 12ℏω in 7Li and 10B nuclei"}
+        ],
+    }
+    record = create_record("lit", data=data)
+    record_control_number = record["control_number"]
+
+    expected_status_code = 200
+    expected_result = '@article{637275237,\n    title = "{Low-energy theorem for $\\gamma\\to 3\\pi$: $\Sigma$ surface terms against $\\pi a_1$-mixing}"\n}\n'
```
Comment (michamos) [inline backend/tests/integration/records/serializers/test_bibtex.py]:
```suggestion
    expected_result = '@article{637275237,\n    title = "{Core polarization effects up to 12$\hbar\omega$ in 7Li and 10B nuclei}"\n}\n'
```

File: backend/inspirehep/records/serializers/bibtex.py [hunk 12]
```diff
@@ -16,6 +19,27 @@
 
 
 class BibtexWriter(Writer):
+    def _encode(self, text):
+        import latexcodec  # NOQA
+
+        if "$" not in text:
+            return codecs.encode(text, "ulatex+{}".format(self.encoding))
+        else:
```
Comment (michamos) [inline backend/inspirehep/records/serializers/bibtex.py]: also [flat is better than nested](https://www.python.org/dev/peps/pep-0020/): you don't need the else here, as the function `return`s if it takes the `if` branch.

File: backend/inspirehep/records/serializers/bibtex.py [hunk 13]
```diff
@@ -16,6 +19,31 @@
 
 
 class BibtexWriter(Writer):
+    math_expression_regex = re.compile(r"(?<!\\)\$.*?\$")
+
+    def _encode(self, text):
+        import latexcodec  # NOQA
```
Comment (michamos) [inline backend/inspirehep/records/serializers/bibtex.py]: you can also move the import out of the method.

File: backend/inspirehep/records/serializers/bibtex.py [hunk 14]
```diff
@@ -16,6 +20,25 @@
 
 
 class BibtexWriter(Writer):
+    math_expression_regex = re.compile(r"(?<!\\)\$.*?\$")
+
+    def _encode(self, text):
+        if "$" in text:
```
Comment (michamos) [inline backend/inspirehep/records/serializers/bibtex.py]: That's not what I meant :sweat_smile: 
```
if "$" not in text:
    return ...
# rest of the function body without else
```

File: backend/tests/integration/records/serializers/test_bibtex.py [hunk 15]
```diff
@@ -139,3 +142,54 @@ def test_bibtex_search(api_client, db, es, create_record):
...
     assert expected_result_2 in response_data
+
+
+def test_bibtex_doesnt_encode_math_environments(api_client, db, es, create_record):
+    headers = {"Accept": "application/x-bibtex"}
+    data = {
+        "control_number": 637275237,
+        "titles": [
+            {
+                "title": "Low-energy theorem for $\\gamma\\to 3\\pi$: $\Sigma$ surface terms against $\\pi a_1$-mixing"
```
Comment (michamos) [inline backend/tests/integration/records/serializers/test_bibtex.py]:
```suggestion
                "title": "Low-energy theorem for $\\gamma\\to 3\\pi$: Σ surface terms against $\\pi a_1$-mixing"
```
The test needs to assert that if you have a unicode character, it gets encoded properly:
```
In [1]: import latexcodec                                                       

In [2]: import codecs                                                           

In [3]: codecs.encode("Σ", "ulatex")                                            
Out[3]: '$\\Sigma$'
```
If that doesn't happen, there's a bug in your logic, which this test is meant to prevent.

File: backend/inspirehep/records/serializers/bibtex.py (hunk 13 above)
Comment (michamos) [reply backend/inspirehep/records/serializers/bibtex.py]: you didn't move it, you copied it.

File: backend/inspirehep/records/serializers/bibtex.py [hunk 16]
```diff
@@ -4,6 +4,10 @@
 #
 # inspirehep is free software; you can redistribute it and/or modify it under
 # the terms of the MIT License; see LICENSE file for more details.
+import codecs
+import re
+
+import latexcodec
```
Comment (michamos) [inline backend/inspirehep/records/serializers/bibtex.py]: you need to silence `flake8` here (http://flake8.pycqa.org/en/3.1.1/user/ignoring-errors.html#in-line-ignoring-errors).

File: backend/inspirehep/records/serializers/bibtex.py [hunk 17]
```diff
@@ -16,6 +20,25 @@
...
 
 class BibtexWriter(Writer):
+    math_expression_regex = re.compile(r"(?<!\\)\$.*?\$")
+
+    def _encode(self, text):
+        import latexcodec  # NOQA
+
+        if "$" not in text:
+            return codecs.encode(text, "ulatex+{}".format(self.encoding))
+        math_expressions = BibtexWriter.math_expression_regex.findall(text)
```
Comment (michamos) [inline backend/inspirehep/records/serializers/bibtex.py]: attributes and methods on the current class can be accessed through `self`:

```suggestion
        math_expressions = self.math_expression_regex.findall(text)
```

File: backend/inspirehep/records/serializers/bibtex.py [hunk 18]
```diff
@@ -16,6 +20,25 @@
...
+        encoded_text = codecs.encode(text, "ulatex+{}".format(self.encoding))
+        encoded_math_expressions = BibtexWriter.math_expression_regex.findall(
```
Comment (michamos) [inline backend/inspirehep/records/serializers/bibtex.py]:
```suggestion
        encoded_math_expressions = self.math_expression_regex.findall(
```
